src/query/sharing/src/signer.rs

```diff
@@ -0,0 +1,231 @@
+use http::Request;
+use moka::sync::Cache;
+use opendal::http_util::AsyncBody;
+use opendal::http_util::HttpClient;
+use opendal::ops::Operation;
+use opendal::ops::PresignedRequest;
+
+/// SharedSinger is used to track presign request, and it's response.
+///
+/// There is an internal cache about presign request. Getting an expired
+/// request will get `None`. Please sign it again.
+#[derive(Clone)]
+pub struct SharedSinger {
+    endpoint: String,
+    token: String,
+    cache: Cache<PresignRequest, PresignedRequest>,
+    client: HttpClient,
+}
+
+impl Debug for SharedSinger {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("SharedSinger")
+            .field("endpoint", &self.endpoint)
+            .finish_non_exhaustive()
+    }
+}
+
+impl SharedSinger {
+    /// Create a new SharedSinger.
+    pub fn new(endpoint: &str, token: &str) -> Self {
+        let cache = Cache::builder()
+            // Databend Cloud Presign will expire after 3600s (1 hour).
+            // We will expire them 10 minutes before to avoid edge cases.
+            .time_to_live(Duration::from_secs(3000))
+            .build();
+
+        Self {
+            endpoint: endpoint.to_string(),
+            token: token.to_string(),
+            cache,
+            client: HttpClient::new(),
+        }
+    }
+
+    /// Get a presign request.
+    pub fn get(&self, path: &str, op: Operation) -> Option<PresignedRequest> {
+        self.cache.get(&PresignRequest {
+            path: path.to_string(),
+            op,
+        })
+    }
+
+    /// Fetch a presigned request. If not found, build a new one by sign.
+    pub async fn fetch(&self, path: &str, op: Operation) -> Result<PresignedRequest> {
+        match self.get(path, op) {
+            Some(v) => Ok(v),
+            None => {
+                self.sign(PresignRequest::new(path, op)).await?;
+                self.get(path, op)
+                    .ok_or_else(|| anyhow!("presigned request should exist, but not"))
+            }
+        }
+    }
+
+    /// Set a presigned request.
+    ///
+    /// This operation will update the expiry time about this request.
+    pub fn set(&self, path: &str, op: Operation, signed: PresignedRequest) {
+        self.cache.insert(PresignRequest::new(path, op), signed)
+    }
+
+    /// Sign a reqeust.
+    pub async fn sign(&self, req: PresignRequest) -> Result<()> {
+        self.sign_inner(vec![req]).await
+    }
+
+    /// Batch sign multiple requests at once.
+    pub async fn batch_sign(&self, reqs: Vec<PresignRequest>) -> Result<()> {
+        self.sign_inner(reqs).await
+    }
+
+    /// Databend Cloud Presign API will be:
+    ///
+    /// ```shell
+    /// curl -X POST \
+    ///      https://endpoint/tenant/<tenant_id>/database/<db_id>/table/<table_id>/presign \
+    ///      --header 'Authorization Bearer OIDC TOKEN' \
+    ///      --data-raw '[{"path": "file_a", "method": "GET"}, {"path": "file_b", "method": "PUT"}'
```

Comment:
I think this design make sense, as we could add additonal fields such as customized headers, expiration time in the PresignRequestItem with backward compatibility